## Code Review

Collapses `Storage_base::at<T>`/`back<T>`/`data<T>` from 11 hand-written specializations each into single template bodies. The refactor itself is correct and well-executed — but it introduces a **downstream-compat regression** that must be fixed before merge.

### The refactor is correct (verified in full)
- Single template bodies; the dtype check now comes from `cy_typeid_v<T>` — one source of truth, so the #965-style per-type drift cannot recur. ✅
- `type_spelling<T>` (variable template with a `static_assert` default) reproduces the old inline spellings, and the new bodies are byte-faithful to the old ones (bounds check on `at`, empty check on `back` — even the "stoarge" typo preserved — `cudaDeviceSynchronize`, `static_cast`). The `"...try to get %s type..."` + `type_spelling<T>` renders byte-identical. ✅
- Concept-constrained declarations + an unconstrained `= delete` fallback → an unsupported `T` is now a clean **compile-time** error at the call site (was a link error). GPU `data<cuDoubleComplex/cuFloatComplex>` specializations are `#ifdef UNI_GPU`-guarded and header-declared (ODR-correct); `StorageDataType` is correctly used on the `Storage` wrapper's `data<T>` to admit them. ✅
- Explicit instantiations preserve all exported symbols (the `nm` diff in the description confirms). ✅

### 🔴 Blocker — breaks `DownstreamFindPackage` (CI-verified)
```
include/Type.hpp:194: error: 'concept' does not name a type; did you mean 'const'?
include/backend/Storage.hpp:55: error: 'CytnxType' has not been declared
```
This PR introduces the first C++20 `concept`s into a universally-included public header (`Type.hpp`), but cytnx sets C++20 only via the build-level `CMAKE_CXX_STANDARD` variable — it does **not** export the requirement to consumers. So `find_package(cytnx)` consumers compile the headers in their default (pre-C++20) mode and fail to parse `concept`. `DownstreamFindPackage` was green on the recent merges (#972/#977/#998), so this is a regression this PR surfaces.

**Fix:** have cytnx declare its language requirement to consumers on the installed/exported target, e.g.
```cmake
target_compile_features(cytnx PUBLIC cxx_std_20)
```
rather than relying only on the build-level `CMAKE_CXX_STANDARD`. That's a CMake change not in this PR. Silver lining: it fixes a genuine latent gap — the public headers already require C++20 (`std::format`, `std::source_location`); the `concept` keyword just makes it undeniable, and exporting the requirement benefits every downstream consumer.

### Notes
- Behind master but `MERGEABLE`; `BuildAndTest` is green on Linux + macOS (cytnx's own build is C++20). macOS-15-intel + wheels were still pending.

**Recommendation: request changes** — the dedup + API hardening is good and should land; it just needs the `PUBLIC cxx_std_20` export (or equivalent) so `DownstreamFindPackage` passes.

*Posted by Claude Code on behalf of @pcchen*

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -299,6 +299,12 @@ add_library(cytnx STATIC)
 set_property(TARGET cytnx PROPERTY C_VISIBILITY_PRESET hidden)
 set_property(TARGET cytnx PROPERTY VISIBILITY_INLINES_HIDDEN ON)
 
+# The public headers require C++20 (std::format, std::source_location, and the
+# CytnxType concept in Type.hpp). Export that requirement through the installed/
+# exported target so find_package(cytnx) consumers compile the headers in C++20
+# mode -- the build-level CMAKE_CXX_STANDARD does not propagate to consumers.
+target_compile_features(cytnx PUBLIC cxx_std_20)
+
 target_include_directories(cytnx
   PRIVATE
   ${CMAKE_CURRENT_SOURCE_DIR}/src

